File: src/brs/util/Listeners.java
package brs.util;

import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;

public final class Listeners<T,E extends Enum<E>> {
  private final ConcurrentHashMap<Enum<E>, List<Consumer<T>>> listenersMap = new ConcurrentHashMap<>();

  public boolean addListener(Consumer<T> listener, Enum<E> eventType) {
    synchronized (this) {
        List<Consumer<T>> listeners = listenersMap.computeIfAbsent(eventType, k -> new CopyOnWriteArrayList<>());
        return listeners.add(listener);
    }
  }

  public boolean removeListener(Consumer<T> listener, Enum<E> eventType) {
    synchronized (this) {
      List<Consumer<T>> listeners = listenersMap.get(eventType);
      if (listeners != null) {
        return listeners.remove(listener);
      }
    }
    return false;
  }

  public void accept(T t, Enum<E> eventType) {
    List<Consumer<T>> listeners = listenersMap.get(eventType);
    if (listeners != null) {
      for (Consumer<T> listener : listeners) {
        listener.accept(t);
      }
    }
  }
}

File: src/main/kotlin/brs/util/Listeners.kt
package brs.util

import brs.util.sync.read
import brs.util.sync.write
import java.util.concurrent.locks.StampedLock

class Listeners<T, E : Enum<E>> {
    private val listenersMap = mutableMapOf<Enum<E>, MutableList<(T) -> Unit>>()
    private val stampedLock = StampedLock()

    fun addListener(eventType: Enum<E>, listener: (T) -> Unit) {
        stampedLock.write {
            listenersMap.computeIfAbsent(eventType) { mutableListOf() }.add(listener)
        }
    }

    fun accept(eventType: Enum<E>, t: T) {
        // Read via the stamped lock from the map, return if null, otherwise each listener should accept the value.
        (stampedLock.read { listenersMap[eventType] } ?: return).forEach { it(t) }
    }
}
